### DISCUSSION TITLE
Hidden control flow with computed properties

### DISCUSSION THREAD

**Discussion by mrael2** [hylo-lang/hylo#1192] [Language design]
Perhaps my opinion is in the minority, but I've always thought that computed properties were a little odd. They behave much like a function/method in Objective-C and Swift, and I see in the Hylo language tour that they are a no-parameter subscript which will yield a value.

But when I'm reading code that has some dot notation without parentheses (or square brackets), I'm inclined to think that I'm dealing with a regular old property and its value, not a function call that might involve multiple statements being executed.

The folks working on the Zig language argue that no hidden control flow is a big, important feature that differentiates their systems programming language from others, and I'm decidedly in agreement with them.

I do like many things about Hylo so far, especially its claims of simplicity, but is there a chance that the language could be further simplified by removing computed properties to remove the hidden control flow found therein?

I'm willing to be educated about why computed properties are absolutely necessary to realize the goals of mutable value semantics if they would indeed play such a role.

**Comment by dabrahams** [comment]
They're not absolutely necessary.  If we didn't have them, some places currently spelled `foo.bar` would be spelled `foo.bar[]`.  

I'm not dead set against the idea.  At the same time, I do believe in abstraction, which means stuff gets hidden, one way or another; I therefore find the claim that no-hidden-control-flow is a huge win suspect.  It's pretty clear to me that forcing people to add `[]` isn't going to improve code in _all_ circumstances.  Hylo is not trying to be a systems-programming-at-the-expense-of-everything-else language. I think one good compromise position might be to keep computed properties, but make trailing `[]` optional by default and provide a command-line option that requires them.

That said, there's very little Hylo code in existence right now, so we don't have a good way to assess the damage, if any, that might be caused by removing computed properties altogether, and I'm not certain they (or parameter-less subscripts) would be common in _any_ code.  My usual thought is to start conservatively by removing features and waiting for a real need to be demonstrated, and my suggested compromise position is pretty messy.

So, color me in favor of running an experiment where we remove computed properties and see what happens.

**Comment by kyouko-taiga** [comment]
In general, I don't think we can promise that accessing a value via dot-syntax is always as cheap as computing a pointer offset. That would be a contract very difficult to honor in a generic context.

Take a look at this API:

```
public trait Identifiable {
  property identity: Int { get }
}

public fun print_id<T: Identifiable>(_ x: T) { print(x.identity) }
```

Unless we can guarantee that this function is always monomorphized, there's no way to promise that accessing `x.identity` is just a pointer offset. That's because if we need to call it across a module boundary, we'll have to wrap the `Identifiable` instance into an existential and perform dynamic dispatch.

We could say that traits are not allowed to define property requirements, but that sounds like a frustrating restriction to me. I also think that it would impede on our ability to apply generic programming without introducing boilerplate when we see that two types share common properties. For example, say you're looking to refactor this code:

```
pubic type A: Regular {
  let identity: Int
  // some other stuff
}

pubic type B: Regular {
  let identity: Int
  // some other stuff
}

fun log(_ x: Union<A, B>) {
  match x {
    let a: A { print(a.identity) }
    let b: B { print(b.identity) }
  }
}
```

It looks quite tempting to say that both `A` and `B` are `Identifiable`, conclude that `Union<A, B>` is `Identifiable` too, and then use `print_id` in `log`. We can do that by just adding `: Identifiable` to the type declarations. If traits can't define properties, however, then we have to write functions or parameter-less subscripts that just reproduce the existing non-generic API.

That being said, I agree that it should probably be considered bad practice to have computed properties with non-constant time complexity. Such a rule would be difficult to enforce but we could make the effort to guide the community to exercise self-discipline.

**Reply by dabrahams** [reply]
> In general, I don't think we can promise that accessing a value via dot-syntax is always as cheap as computing a pointer offset.

Sure we can. A trait author who wants to allow for a more expensive operation would just have to require a subscript instead of a property. 

I'm not saying we _want_ that world, but it's one we can have.

**Reply by kyouko-taiga** [reply]
Hmm, I'm not sure.

```
trait P {
  property x: Int { get }
}

type A: Regular, P {
  public memberwise init
  public let x: Int
}

// So far, so good but ...

fun recurse_randomly<T: P, U>(_ t: T, _ u: U) -> Int {
  if Bool.random() {
    recurse_randomly(t, (u, ()))
  } else {
    t.x.copy() // not a pointer offset computation
  }
}

public fun main() {
  let a = A()
  _ = recurse_randomly(a, ())
}
```

The problem is that `recurse_randomly(_:_:)` can't be monomorphized, so we'll have to create an existential around `a`. This existential is accompanied with a witness table that tells the runtime how to access `x` and using that witness table is more expensive than a pointer offset. There's nothing the trait author could say to avoid this situation.

**Comment by tothambrus11** [comment]
Computed properties are great for hiding implementation details. That's why people opt in to use the verbose getters and setters in Java and C++. Java has Lombok, a dynamic reflection system to autogenerate getters and setters, and I guess C++ will have to use static reflection for autogenerating these. Swift/Hylo properties are great, because their representation can be changed at any time without breaking users of the API.

**Comment by tothambrus11** [comment]
A use case for computed properties is the C interoperability scheme discussed in [this paper](https://ambrus.dev/_astro/High-Fidelity%20C%20Interoperability%20in%20Hylo.7fUYiq3A.pdf). A struct from C, with potentially incompatible memory layout to Hylo can be exposed as a Hylo struct that stores a private byte array. All the C properties are then exposed to Hylo via computed properties, that offset this byte array and do some type punning and potential data transformation if needed. This works perfectly for both reading, writing and modification, and it's likely very little overhead at runtime.

**Comment by tothambrus11** [comment]
Btw I don't think computed properties hide any control flow in any way that the caller should care about. I would contrast this to function calls of throwing functions, which in C++ do hide control flow. Swift uses explicit `try` annotations on potentially throwing function calls, which makes that more explicit.

Anyways, I do aggree that we shouldn't put expensive computation into computed properties. E.g. `count` on `LinkedList` should be a function rather than a property, and for that reason, the requiremeint in the `Collection` /`Sequence` API should be also a function instead of a property.

**Comment by kyouko-taiga** [comment]
I am no longer convinced that computed properties are such a great idea, in part because of the points OP made about hidden costs. In addition, computed properties create friction during type inference because the compiler has to figure out that the abstraction needs to be applied, which is not obvious in a type system. The current implementation does not handle that in a principled way and that is the source of several issues IR lowering.

I don't think C interop. requires computed properties because I don't see a problem writing an extension to any blob of data that present a subscript with no arguments.

**Reply by dabrahams** [reply]
But resilience means that all properties are computed, in general. Unless you want to say there are no resilient properties. I don't see how being computed or not can have any impact on inference, because the declaration of a computed property is the same as the declaration of a non-computed one, from the client side.
